### PR TITLE
feat: add scope information to the ast

### LINKED ISSUE #120: add scope information to the AST
in `no-unused-vars` (#22) I added some information to the AST, which is not ideal to do in a rule because the order of execution of the different rules is undefined (as in, not part of any API) and one rule could inadvertedly break others.

The DoD of this issue is to
- [ ] have the information regarding identifier scoping included in the AST in
  a step that's previous to rule execution, similar to how we use `ast-parents`
- [ ] make `no-unused-vars` a smaller rule that doesn't modify the AST
- [ ] consider the creation of a `no-shadowing` rule reaping the benefits of
  the above
- [ ] check that no other rules modify the AST, create issues if they do

### PR DESCRIPTION
**Pull request by jahabeebs**
Closes #120 

Here is a brief summary of what I've done for adding scope info to the AST
1) Created scope-manager.js which handles the lexical scoping + tracking logic to prevent no-unused-vars from needing to modify the AST

Scope-manager.js has 100% test code coverage. However, I would recommend some performance testing before the release by checking with some repos with large amount of Solidity examples to find edge cases we haven't considered. Ideally for future work we would do an incremental migration of solhint to TypeScript, starting with scope-manager.js, to ensure that we not only cover 100% of the code but also all of the possible branches and ensure there are no edge cases that could cause an error and/or redundant guards. Note that in the JSDocs I've included the types of each function which should help with an incremental migration because these should be the accurate types (though currently they're not enforced)

2) Made no-unused-vars a smaller rule that does not modify the AST--we have 100% code coverage for no-unused-vars.js as well but I'd make the same recommendations as above

Future work:

- check that no other rules modify the AST

code-complexity.js and reentrancy.js at least

- no-shadowing rule https://github.com/solhint-community/solhint-community/issues/156

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,13 @@
+## [4.1.0] - XXXX-XX-XX
+
+### Added
+- introduced `scope-manager.js` to handle lexical scoping and variable tracking prior to rule execution
```
Comment (juanpcapurro) [inline CHANGELOG.md:4]: since this is mostly meant as an overview meant for _users_ of the linter which are not likely to care about how we organize files, I'd say to keep the update to a single item under `###Refactored` or something similar that makes it clear this does not modify the api/behaviour.

File: lib/common/scope-manager.js [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+  getScope(node) {
+    return this.scopes.get(this.findScopeNode(node))
+  }
+
+  /**
+   * Traverses up the AST to find the nearest ancestor node that has a scope
+   * @param {ASTNode} node - Starting node to search from
+   * @returns {ASTNode|undefined} The nearest ancestor with a scope, or undefined if none exists
+   */
+  findScopeNode(node) {
```
Comment (juanpcapurro) [inline lib/common/scope-manager.js:41]: have you considered using `findFirstParent` from `lib/common/tree-traversing`?

File: lib/common/scope-manager.js [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+/**
+ * Checks if an identifier is being used in its own import declaration
+ * (e.g. the 'A' in 'import {A} from "./A.sol"')
+ * @param {ASTNode} idNode - Identifier node
+ * @returns {boolean} True if identifier is part of its own import
+ */
+function isImportDirectiveSelfReference(idNode) {
+  let cur = idNode.parent
+  while (cur) {
+    if (cur.type === 'ImportDirective') {
```
Comment (juanpcapurro) [inline lib/common/scope-manager.js:144]: have you taken into consideration [this issue](https://github.com/solhint-community/solhint-community/issues/119) ?

File: lib/common/scope-manager.js [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+    SourceUnit(node) {
+      scopeManager.addScope(node)
+    },
+
+    ContractDefinition(node) {
+      scopeManager.addScope(node)
+    },
+
+    FunctionDefinition(node) {
+      if (!hasNoBody(node)) {
```
Comment (juanpcapurro) [inline lib/common/scope-manager.js:217]: I found this double negation confusing, I suggest to avoid function names with negations in them, and use the unary `!` if applicable

File: lib/common/scope-manager.js [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+          while (sc) {
+            if (sc.trackUsage(id.name)) break
+            const p = sc.node?.parent
+            sc = p ? scopeManager.getScope(p) : null
+          }
+        }
+      }
+
+      // Recursively visit assembly nodes to track identifier usage
+      const visit = (n) => {
```
Comment (juanpcapurro) [inline lib/common/scope-manager.js:261]: why do this instead of visiting `AssemblyLocalDefinition`, `AssemblyAssignment`, etc? does the default visitor not get you  through them?

File: lib/common/scope-manager.js [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+        }
+      }
+    },
+
+    // Ignore state variables in unused checks
+    StateVariableDeclaration(node) {
+      const scope = scopeManager.getScope(node)
+      if (!scope) return
+      for (const v of node.variables) {
+        if (v && v.name) {
```
Comment (juanpcapurro) [inline lib/common/scope-manager.js:332]: why wouldn't `v` be defined? Also, be consistent with `?.` usage

File: lib/common/scope-manager.js [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+   * @param {ASTNode} node - The AST node representing the variable declaration
+   * @param {string} name - The name of the variable
+   * @param {Object} opts - Options object
+   * @param {boolean} opts.ignore - If true, variable will be tracked but ignored in unused checks
+   */
+  addVariable(node, name, opts = {}) {
+    if (!name) return
+    // Special case: skip 'payable' as it can be both a type and parameter name
+    if (name === 'payable') return
+    if (opts.ignore) return
```
Comment (juanpcapurro) [inline lib/common/scope-manager.js:71]: why not simply choose to not invoke the function? also, how is it tracked but not taken into consideration for unused checks?

this will get in the way of the `no-shadowing` rule iiuc

File: lib/rules/best-practises/no-unused-vars.js [hunk 8]
```diff
@@ -70,16 +67,16 @@ class NoUnusedVarsChecker extends BaseChecker {
...
+   */
 
   'SourceUnit:exit'(node) {
     // get inheritdoc statements from the lexed (but not parsed) input
-    const keywords = this.tokens.filter((it) => it.type === 'Keyword')
-    const inheritdocStatements = keywords.filter(
-      ({ value }) => /^\/\/\/ *@inheritdoc/.test(value) || /^\/\*\* *@inheritdoc/.test(value)
+    const keywords = this.tokens.filter((t) => t.type === 'Keyword')
+    const inheritdocStatements = keywords.filter((k) =>
+      /^\/\/\/ *@inheritdoc|^\/\*\* *@inheritdoc/.test(k.value)
```
Comment (juanpcapurro) [inline lib/rules/best-practises/no-unused-vars.js:79]: are you sure you don't have to group both sides of the branch? how's the precedence?

File: lib/rules/best-practises/no-unused-vars.js [hunk 9]
```diff
@@ -89,209 +86,93 @@ class NoUnusedVarsChecker extends BaseChecker {
...
 
-    node.scope.unusedVariables().forEach(this._error.bind(this))
-  }
-
-  Block(node) {
-    if (node.parent?.type === 'FunctionDefinition') {
-      return
-    }
-    VarUsageScope.activate(node)
+    this._reportUnused(node)
```
Comment (juanpcapurro) [inline lib/rules/best-practises/no-unused-vars.js:93]: why report the unused variables when _entering_ the block?

File: lib/rules/best-practises/no-unused-vars.js [hunk 10]
```diff
@@ -89,209 +86,93 @@ class NoUnusedVarsChecker extends BaseChecker {
...
-  hasVar(name) {
-    return this.vars[name] !== undefined
-  }
-
-  trackVarUsage(name) {
-    const curVar = this.vars[name]
-
-    if (curVar) {
-      curVar.usage += 1
+    // Traverse up until we find a node with a valid scopeManager
```
Comment (juanpcapurro) [inline lib/rules/best-practises/no-unused-vars.js:167]: why do you search for it every time if you know it is on the root?

File: lib/rules/best-practises/no-unused-vars.js [hunk 11]
```diff
@@ -89,209 +86,93 @@ class NoUnusedVarsChecker extends BaseChecker {
...
-          value: node,
-          configurable: true,
-          enumerable: false,
-          writable: true,
-        })
-        this._addVariable(idNode)
-      })
+  _reportUnused(node) {
+    const scope = this._getScope(node)
+    if (!scope) return
```
Comment (juanpcapurro) [inline lib/rules/best-practises/no-unused-vars.js:138]: all tests continue to pass if i comment this line

File: lib/rules/best-practises/no-unused-vars.js [hunk 12]
```diff
@@ -89,209 +86,93 @@ class NoUnusedVarsChecker extends BaseChecker {
...
+    // Handle assembly usage of parameters
+    const hasAssembly = node.body.statements.some((st) => st.type === 'InlineAssemblyStatement')
+
+    if (hasAssembly) {
+      // Parameters used in assembly are considered used
+      for (const param of node.parameters || []) {
+        if (param.name) {
+          const scope = this._getScope(node)
+          if (scope) {
+            scope.trackUsage(param.name)
```
Comment (juanpcapurro) [inline lib/rules/best-practises/no-unused-vars.js:127]: all tests continue to pass if I comment this line

**Review by juanpcapurro — CHANGES_REQUESTED** [review]
Hey, thanks for taking a stab at this!

I don't want to be the guy who rejects a patch because it's not implemented the way _he_ would do it, however I am puzzled by the way you chose to design this.

Some of the questions I have are throughout the diff, but in a more general sense I'm curious about why did you choose to keep the scope tracking in a separate object at the root of the ast instead of attaching each scope to the relevant ast node associated with it (`SourceUnit`, `FunctionDefinition`, `Block`)? that is analogous to what the `no-unused-var` rule currently does, so it should yield the easiest refactor.

I know some rules are implemented like that, with the objective of not polluting the AST with rule-specific information, but we're trying to do the opposite here, and embed useful info into the ast. I imagine there could be other reasons to have a separate scope-tracking object, and am curious of those as well.
